src/daemon.py

```diff
@@ -161,58 +68,88 @@ def execute_yara(job: JobId, files: List[str]) -> None:
         return
 
     rule = compile_yara(job)
-    len_matches = 0
-    db.set_files_in_progress(job, len(files))
+    num_matches = 0
+    db.job_start_work(job, len(files))
     for sample in files:
         try:
             matches = rule.match(sample)
             if matches:
-                len_matches += 1
+                num_matches += 1
                 update_metadata(job, sample, [r.rule for r in matches])
         except yara.Error:
             logging.exception(f"Yara failed to check file {sample}")
         except FileNotFoundError:
             logging.exception(f"Failed to open file for yara check: {sample}")
-        db.update_files_in_progress(job)
-
-    db.update_job(job, len(files), len_matches)
-    job_data = db.get_job(job)
-    if (
-        job_data.files_in_progress == 0
-        and job_data.files_processed >= job_data.total_files
-    ):
-        db.finish_job(job)
 
+    db.job_update_work(job, len(files), num_matches)
 
-def execute_search(job_id: JobId) -> None:
-    logging.info("Parsing...")
 
-    job = db.get_job(job_id)
-    yara_rule = job.raw_yara
+def do_search(job: JobId, iterator: str):
+    try:
+        BATCH_SIZE = 500
+        pop_result = ursa.pop(iterator, BATCH_SIZE)
+        if pop_result.was_locked:
+            logging.info(
+                "Iterator %s is locked, retrying in a second", iterator
+            )
+            return False
+        if pop_result.files:
+            execute_yara(job, pop_result.files)
+        if pop_result.should_drop_iterator:
+            logging.info(
+                "Dropping job %s because iterator %s is empty",
+                job.key,
+                iterator,
+            )
+            return True
+    except Exception:
+        logging.exception("Failed to execute yara match.")
+        return True
```

Comment:
I think that logic is a bit unusual here: maybe we should return False in case of failure instead of True?

Reply:
*in case of finish